diffrax/brownian/tree.py

```diff
@@ -88,70 +152,224 @@ def __init__(
+            w_st = w_term1 + w_term2
+            w_tu = w_term1 - w_term2
+            w_st_tu = (w_st, w_tu)
+
+            bhh_term1 = bhh_su / 8 - su / 4 * z
+            bhh_term2 = su / 4 * n
+            bhh_st = bhh_term1 + bhh_term2
+            bhh_tu = bhh_term1 - bhh_term2
+            bhh_st_tu = (bhh_st, bhh_tu)
+
+            w_t = w_s + w_st
+            bhh_t = bhh_s + bhh_st + 0.5 * (t * w_s - s * w_t)
+            bhh = (bhh_s, bhh_t, bhh_u)
+            bkk = None
+            bkk_st_tu = None
+
+        else:
+            assert bhh is None
+            assert bkk is None
+            mean = 0.5 * w_su
+            w_term2 = root_su / 2 * jrandom.normal(key, shape, dtype)
+            w_st = mean + w_term2
+            w_tu = mean - w_term2
+            w_st_tu = (w_st, w_tu)
+            w_t = w_s + w_st
+            bhh, bhh_st_tu, bkk, bkk_st_tu = None, None, None, None
+        return t, w_t, w_st_tu, bhh, bhh_st_tu, bkk, bkk_st_tu
 
     def _evaluate_leaf(
         self,
         key,
-        τ: Scalar,
+        r: Scalar,
         shape: jax.ShapeDtypeStruct,
-    ) -> Array:
+    ) -> LevyVal:
         shape, dtype = shape.shape, shape.dtype
 
-        cond = self.t0 < self.t1
-        t0 = jnp.where(cond, self.t0, self.t1).astype(dtype)
-        t1 = jnp.where(cond, self.t1, self.t0).astype(dtype)
+        t0 = jnp.zeros((), dtype)
+        t1 = jnp.ones((), dtype)
+        r = jnp.asarray(r, dtype)
 
-        t0 = eqxi.error_if(
-            t0,
-            τ < t0,
-            "Cannot evaluate VirtualBrownianTree outside of its range [t0, t1].",
-        )
-        t1 = eqxi.error_if(
-            t1,
-            τ > t1,
-            "Cannot evaluate VirtualBrownianTree outside of its range [t0, t1].",
-        )
-        # Clip because otherwise the while loop below won't terminate, and the above
-        # errors are only raised after everything has finished executing.
-        τ = jnp.clip(τ, t0, t1).astype(dtype)
+        w_0 = jnp.zeros(shape, dtype)
+
+        if self.levy_area == "space-time":
+            key, init_key_w, init_key_la, midpoint_key = jrandom.split(key, 4)
+            w_1 = jrandom.normal(init_key_w, shape, dtype)
 
-        key, init_key = jrandom.split(key, 2)
-        thalf = t0 + 0.5 * (t1 - t0)
-        w_t1 = jrandom.normal(init_key, shape, dtype) * jnp.sqrt(t1 - t0)
-        w_thalf = self._brownian_bridge(t0, thalf, t1, 0, w_t1, key, shape, dtype)
+            bhh_1 = jnp.sqrt(1 / 12) * jrandom.normal(init_key_la, shape, dtype)
+            bhh_0 = jnp.zeros_like(bhh_1)
+            bhh = (bhh_0, bhh_1, bhh_1)
+            bkk = None
+
+        else:
+            key, init_key_w, midpoint_key = jrandom.split(key, 3)
+            w_1 = jrandom.normal(init_key_w, shape, dtype)
+            bhh = None
+            bkk = None
+
+        w = (w_0, w_1, w_1)
+
+        half, w_half, w_inc, bhh, bhh_inc, bkk, bkk_inc = self._brownian_arch(
```

Comment:
Heads-up that it's good practice to try and avoid a separate call outside of a loop, if possible: this is because JAX will compile each call to `_brownian_arch` separately. So instead picking differential initial conditions and letting the loop run for one more iteration is preferred.

That said, in this case `_brownian_arch` is pretty small, so if that's awkward to do don't worry about that here, and just treat it as a learning point.

Reply:
Originally I just copied your design choice here. But now that you mention this, I realised it could be improved significantly, so I overhauled it. Now all calls to _brownian_arch are inside the lax.scan. This also makes _State much more concise, and the final interpolation is a bit nicer. I quickly did some tests, and it seems everything works fine, and also somewhat faster.